### DISCUSSION TITLE
Fody and especially Fody.PropertyChanged should be built in

### DISCUSSION THREAD

**Discussion by buzzware** [dotnet/csharplang#4200] [Language Ideas]
### Summary

Fody and especially Fody.PropertyChanged should be built in.

### Epics

Epics under this Theme:

- [ ] [Title of the linked Epic](hyperlink to the Epic issue).

**Comment by scalablecory** [comment]
What is Fody? Can you expand on your suggestion?

**Comment by svick** [comment]
@scalablecory [Fody is an open source IL weaver tool](https://github.com/Fody/Fody) and [PropertyChanged.Fody](https://github.com/Fody/PropertyChanged) is a Fody plugin that lets you avoid boilerplate code that is very common in many apps. (My employer, [PostSharp](https://www.postsharp.net/), has a similar commercial tool with [a similar plugin](https://www.postsharp.net/mvvm).)

Though I think the intention of the C# team is that [source generators](https://devblogs.microsoft.com/dotnet/introducing-c-source-generators/) are the solution for these kinds of problems.

**Reply by JanTamis** [reply]
I think there is even a sample for it.

**Comment by buzzware** [comment]
If you compare Xamarin binding template development with say EmberJs, you'll see how horrible MVVM is without Fody.PropertyChanged. We should not have to manually implement propeties that raise events for every property.

**Comment by scalablecory** [comment]
Thanks @svick. I suspect you're right about source generators 👍

**Reply by buzzware** [reply]
Regarding source generators, it all depends on the developer experience and the tidiness of the models.
People from the Roslyn team should site down with the MVVM framework people and make sure the ideal model definitions are possible. 

I'm thinking from a Xamarin point of view, and I recommend looking at https://mobx-state-tree.js.org/intro/philosophy and https://guides.emberjs.com/release/in-depth-topics/autotracking-in-depth/#toc_tracked-properties-in-custom-classes.

In both State Tree and Ember you can model not just a single entity class but the live state of the entire application in a bindable form including entities and their relationships, all without a line of visual representation. Then you render this state and mutate it with UI actions, and you're done.

In particular, notice the handling of arrays, dependent properties and caching. There is no declaring anything like a BindableProperty.

Normally I avoid source generation because there are issues in keeping the generated source fresh with the actual source. However if that is solved at development time in .Net (and Xamarin already generates source) then I suppose its not an issue.
